Get Device API call

thessbsu opened the pull request:
<!--
These comments inside these brackets will not appear in the pull request.

The pull request should be linked to either:
 - a task (i.e., use 'Closes #TaskID' or 'Resolves #TaskID')
 - a bug (i.e., use 'Closes #BugID' or 'Fixes #BugID')

For more details, see: https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword
-->

Closes #94

This is from my apiCalls.js I used to calibrate the output, just shows usage
const api = require ('./ApiClient');
const returnVar = await api.GetDevice(&lt;device ID&gt;);

Running the api calls and the output. The actual return from the GetDevice call is the JSON object only. The parts after it are showing the members being accessed and what is there.

$ node apiCalls.js

const returnVar= await api.GetDevice(&lt;invalidObjectId&gt;);
{ status: '400', response: 'Invalid id.' }
returnVar.status: 400
returnVar.response: Invalid id.


const returnVar= await api.GetDevice(&lt;validDeviceIdNotInDb&gt;);
{ status: '404', response: 'Device could not be found.' }
returnVar.status: 404
returnVar.response: Device could not be found.


const returnVar= await api.GetDevice(&lt;deviceIdThatCurrentlyExistsInDb&gt;);
{
  status: '200',
  response: [
    { name: '_id', value: '620aeb23f50067dd0535bab3' },
    { name: 'deviceTypeId', value: 'a' },
    { name: 'serialNumber', value: 'a12345' },
    { name: 'updatedDeviceField', value: '123' }
  ]
}
returnVar.status: 200
returnVar.response[0].name: _id
returnVar.response[0].value: 620aeb23f50067dd0535bab3
etc..

josephmoore99 commented:
I don't think this will affect what you have here, but I have updated the format for the json response for the GET endpoint. It should now be in a much more jsony format. It should just have the properties and their values now rather than having the "name" "value" tuples.

josephmoore99 commented on BEIMA.Client/services/ApiClient.js:
```diff
@@ -0,0 +1,32 @@
+
+const API_URL = process.env.REACT_APP_API_URL;
+
+/// <summary>
```
I know this is our header formatting for our C# code, but do we want to use the same formatting here?

thessbsu replied:
All the examples I see after looking to see if there's a standard appear to be using the JSDoc format.
https://en.wikipedia.org/wiki/JSDoc

I'll swap it for that since it seems to be nearly ubiquitous

AshlynAdamson commented on BEIMA.Client/services/ApiClient.js:
```diff
@@ -0,0 +1,32 @@
+//module.exports.<Custom Name> = <Function Name>
+//OR
+//module.exports.<Function Name> = <Function Name>
+module.exports.GetDevice = GetDevice;
```
I think `export default GetDevice` will work here too. That's what we've been following for the other JS code.

thessbsu replied:
Would that interfere with other functions being included in the file?

AshlynAdamson replied:
There's currently only one function in this file, and it might be a good idea to keep each API call in its own file, like how the backend is doing. But if we want to include all the functions in one file, then you can add `export default` in front of the function declaration

Kenny-Miller replied:
You can just put export infront of const GetDevice... if you are wanting to have multiple exported functions in the file. I think the module.exports you have also works just as well.

thessbsu replied:
I tried a few different ways of exporting and a few complained when trying to call them externally, I'll try the export in front of the function name and see how it likes it, I can't remember if I tried that or not.

thessbsu replied:
One per file would limit the amount of external code that gets loaded per page, and the "show one device" doesn't need to know about device/device type list functionality and vice versa (or any other operation).

I'll swap it tonight

AshlynAdamson commented:
I had envisioned these JS connection code tasks to have a bit more with them. The idea was to get a static UI up, and then these tasks would make functions to call the endpoint, like you've done here, but also use them in the UI, making that connection from the backend to the frontend. That is why I have mentioned before that I think these tasks needed to wait for the UI to be done.

Maybe I am the only one that interpreted it this way during sprint planning, and if that's the case then we can role with this and we will have to make another story for next sprint that updates the UI with these calls since story #79 is for initial pages with no functionality to them yet.

thessbsu commented:
> I don't think this will affect what you have here, but I have updated the format for the json response for the GET endpoint. It should now be in a much more jsony format. It should just have the properties and their values now rather than having the "name" "value" tuples.

It shouldn't affect it, I just rewrap it with the return code. The only change I can see is not having to use array indexing to access the members since they'll be flat.

thessbsu commented:
> 

I think I get what you're saying.
The API client should return a React component to be rendered?

I was thinking "this needs to get the data so the calling page can add it to the fields" rather than "return a render-able piece of a page"

If it needs to return a React component, I can do that, it will take a bit of reading and experimenting to get it right.

AshlynAdamson commented:
> I think I get what you're saying. The API client should return a React component to be rendered?
> 
> I was thinking "this needs to get the data so the calling page can add it to the fields" rather than "return a render-able piece of a page"
> 
> If it needs to return a React component, I can do that, it will take a bit of reading and experimenting to get it right.
> 
> If the calling page will be adding data to the fields I have the code for using them outside the module that should be usable in a page. The api client would need to be in a requires directive then, inside a function in the page would have:
> 
> const retVar = await api.GetDevice(<param>);
> 
> then the data itself is in the retVar to be assigned to the fields

I don't think we need to return a render React item, returning the data how you have it now should be fine. What I'm referring to is updating the calling page to actually call this function and verify the page can is updated correctly.

thessbsu commented:
> I don't think we need to return a render React item, returning the data how you have it now should be fine. What I'm referring to is updating the calling page to actually call this function and verify the page is updated correctly.

When thinking it might need to be a component I searched around and found a use case for something else, but I thought in the React component it would look something like this:

import React from 'react';
const api = require(&lt;relative path to ApiClient&gt;);

function ReactFunction() {
  
  function FillDevice(deviceId) {
    const getMyData = await api.GetDevice(deviceId);
	
	// fill device fields or pass the data to another page etc
       "set visible field's value/text to" getMyData.id;
  }
  
  return (
    &lt;ul&gt;
        &lt;li&gt;
          &lt;button onClick={() =&gt; FillMyDeviceFields(deviceId)}&gt;
          &lt;/button&gt;
        &lt;/li&gt;
    &lt;/ul&gt;
  );
}

Kenny-Miller reviewed (changes requested):
I think it would be better if you just returned this

const res = {
   status: response.status,
   response: response.data
}
return res

Kenny-Miller reviewed (changes requested):
If you await, you don't need to use the .then statement. You just would wrap it in a try catch to handle errors. Likewise using the .then you don't need the await. https://stackabuse.com/making-asynchronous-http-requests-in-javascript-with-axios/

AshlynAdamson commented:
> import React from 'react'; const api = require(<relative path to ApiClient>);
> 
> function ReactFunction() {
> 
> function FillDevice(deviceId) { const getMyData = await api.GetDevice(deviceId);
> 
> ```
> // fill device fields or pass the data to another page etc
>    "set visible field's value/text to" getMyData.id;
> ```
> 
> }
> 
This is what I'm referring to. The pages are currently being developed with static device fields, so they already have rendered components for that. They would just need to include a function to dynamically populate the device fields using the API call function you've created, instead of using a static list of data.

Kenny-Miller commented:
> Currently, the pages don't know what those API functions are, that's why I think adding it to the pages needs to be done when we create the API function, therefore we should wait for the UI to be done. This PR can stay up, but once we get the UI done we should update the PR to incorporate the API calls into the page. Maybe we should have a discussion this evening about this.

Its fine doing both at the same time and merging the pr's in so long as we through in a task to merge them together. The ui should be using mock data to base itself off anyways while creating the view. That way the switch over will be as easy as just pasting a function call.

thessbsu commented:
> If you await, you don't need to use the .then statement. You just would wrap it in a try catch to handle errors. Likewise using the .then you don't need the await. https://stackabuse.com/making-asynchronous-http-requests-in-javascript-with-axios/

OH! Yes, it's likely still there as a remnant from when I did do some actions in the Then portion.

AshlynAdamson commented:
> Its fine doing both at the same time and merging the pr's in so long as we through in a task to merge them together. The ui should be using mock data to base itself off anyways while creating the view. That way the switch over will be as easy as just pasting a function call. That's not to say the objects won't change, but that more so depends on the backend endpoints and what they return.

The scope of the UI tasks is to create initial pages, the actual functionality of it is out of scope for those tasks. We have quite a bit of WIP and you and I have been working on those tasks for longer than 1-2 working sessions. We should get the initial UI tasks closed first, then move on to adding the API calls to the pages as we get those functions created. 

If we wait and do them at the same time, it adds way more decencies on those tasks than we had originally had, which is not good. It will take longer for them to close and we need to keep tasks moving through the pipelines rather than having them sit in "In Progress" for pretty much the whole sprint.

thessbsu commented:
doc style was updated to JSDoc format
renamed to indicate single use file for each function that will be implemented
moved to the src/services directory so the pages can access it

The axios catch block was readded since internally axios attaches the DB response and makes it available to be in the returned response, otherwise the return is:
Error: Request failed with status code 404
    at createError (createError.js:16:1)
    at settle (settle.js:17:1)
    at XMLHttpRequest.onloadend (xhr.js:54:1)

without the await the browser catches the error response and puts this in the console and doesn't even get back to my code
GET http://localhost:7071/api/device/621b99ef97238d8b7acdf40e 404 (Not Found)

josephmoore99 commented on BEIMA.Client/services/ApiClient.js:
```diff
@@ -0,0 +1,24 @@
+import axios from 'axios';
```
It looks like you have the same code in three separate files, is this intentional?

thessbsu replied:
It might not be picking up that the original "apiclient" file was renamed to "GetDevice", and then "GetDevice" was moved from /services to /src/services.  The remove of apiclient and move of getdevice might not have been added to the commit. I'll have to check once I'm off work.

thessbsu replied:
Just deleted them through the web interface, one file, where it's supposed to be, is included
